include/spdlog/sinks/qt_sinks.h

```diff
@@ -162,15 +162,23 @@ class qt_color_sink : public base_sink<Mutex>
 
         const string_view_t str = string_view_t(formatted.data(), formatted.size());
         // apply the color to the color range in the formatted message.
-        auto payload = QString::fromLatin1(str.data(), static_cast<int>(str.size()));
+        auto payload = QString::fromUtf8(str.data(), static_cast<int>(str.size()));
+        // convert color ranges from byte index to character index.
+        int color_range_start, color_range_end;
+        color_range_start = static_cast<int>(msg.color_range_start);
+        color_range_end = static_cast<int>(msg.color_range_end);
+        if (msg.color_range_start < msg.color_range_end) {
+            color_range_start = QString::fromUtf8(str.data(), msg.color_range_start).size();
+            color_range_end = QString::fromUtf8(str.data(), msg.color_range_end).size();
```

Comment:
Why use `QString`?
I think you can use `QStringRef` here.

See: https://stackoverflow.com/questions/62910966/how-to-use-qstringview-or-qstringref-in-qt-widgets

Reply:
To convert byte index/number to character index/number, there must be an API that takes byte index/number as input and returns character index/number. This is what `QString::fromUtf8(const char *str, qsizetype size)` and `QString::size()` can accomplish. But I don't see any `QStringRef` API takes byte index/number as input. Or I missed something.

As for `QStringView`, it assumes the string is UTF16-encoded. So not suitable here.

I also looked for `QUtf8StringView`. But can't find any API returns character index/number. (I thought it should return character index/number, but the following test code showed me wrong.)

```C++
    QString test = "测试字符串";
    qDebug() << test.size();
    QByteArray bytes = test.toUtf8();
    QUtf8StringView testView(bytes.data(), bytes.size());
    qDebug() << testView.size();
```

Reply:
I think we need to add  to the ctor a bool flag to indicate if the log is utf8 or not(default false). This way all those expensive conversions can be avoided if the utf8 option is false.

Reply:
I don't know what version of Qt we should support, can't we use the `QByteArrayView` that was added in Q

Reply:
Memory consumption is proportional to the length of log messages.
We want to avoid memory copying just to find the index of a character as much as possible.

Reply:
I find an efficient way for counting character number in UTF8 encoded string. (See: https://stackoverflow.com/questions/4063146/getting-the-actual-length-of-a-utf-8-encoded-stdstring)

Do you think we can replace `QString::fromUtf8` and `QString::size` with this code?

Reply:
If you know the character to look for, you can use `QString::indexOf()`.

https://doc.qt.io/qt-6/qstring.html#indexOf-1

Reply:
But how can I know which one is what I want if there are multiple occurrences of this character?

Reply:
If a user added the same characters as the color code to a message, it would not be distinguishable from the color code generated by the spdlog format.

BTW, I don`t understand why you don`t use [`QStringRef::QStringRef(const [QString](https://doc.qt.io/qt-6/qstring.html) *string, int position, int length)`](https://doc.qt.io/qt-5/qstringref.html#QStringRef-1) (I suggested it once).
Are you saying that the following code does not work?

```cpp
color_range_start = QStringRef(&str msg.color_range_start).size();
color_range_end = QStringRef(&str, msg.color_range_end).size();
```

Reply:
It seems that the API supported by Qt is incompatible with UTF-8.
It seems that `QString::fromUtf8()` has to be used.
Thank you for your time.